File: frontend/src/lib/geo.ts
/**
 * Client-side geo helpers — great-circle distance + display formatting.
 *
 * Mirrors the backend's `apps/common/geo.haversine_km` so "distance to the bus" reads the
 * same on both sides. Straight-line (as-the-crow-flies), not road distance — fine for a
 * "how far is my bus" cue.
 */

export interface LatLng {
  lat: number;
  lng: number;
}

const EARTH_RADIUS_KM = 6371.0088;

const toRad = (deg: number): number => (deg * Math.PI) / 180;

/** Great-circle distance between two points, in kilometres. */
export function haversineKm(a: LatLng, b: LatLng): number {
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(h));
}

/** Human distance: metres under 1 km (`350 m`), else one-decimal km (`1.2 km`). */
export function formatDistance(km: number): string {
  if (!Number.isFinite(km)) return "";
  if (km < 1) return `${Math.round(km * 1000)} m`;
  return `${km.toFixed(1)} km`;
}

File: backend/apps/common/geo.py
"""Small, dependency-free geo helpers shared across apps.

``haversine_km`` is the true great-circle distance between two lat/lng points — the
complement to ``BusStopRepository.nearby``'s bounding-box filter (which trades exactness
for an index-friendly range query). Inputs may be ``float``, ``str``, or ``Decimal`` (GPS
and stop coordinates are ``DecimalField`` on the wire); everything is coerced to float for
the trig. No PostGIS, so this runs identically on SQLite (tests) and Postgres (prod).
"""

from math import asin, cos, radians, sin, sqrt

EARTH_RADIUS_KM = 6371.0088  # mean Earth radius (IUGG)


def haversine_km(lat1, lng1, lat2, lng2) -> float:
    """Great-circle distance between two points in kilometres.

    Accepts numeric or string/Decimal coordinates. Returns 0.0 for identical points.
    """
    lat1_f, lng1_f, lat2_f, lng2_f = (float(lat1), float(lng1), float(lat2), float(lng2))
    d_lat = radians(lat2_f - lat1_f)
    d_lng = radians(lng2_f - lng1_f)
    a = sin(d_lat / 2) ** 2 + cos(radians(lat1_f)) * cos(radians(lat2_f)) * sin(d_lng / 2) ** 2
    return 2 * EARTH_RADIUS_KM * asin(sqrt(a))
